addon/-private/system/store.js

```diff
@@ -1249,25 +1271,23 @@ const Store = Service.extend({
     );
 
     let normalizedModelName = normalizeModelName(modelName);
-
     let trueId = coerceId(id);
-    let internalModel = this._internalModelsFor(normalizedModelName).get(trueId);
+    let identifier = recordIdentifierFor(this, { type: normalizedModelName, id: trueId });
+    let internalModel = this._internalModelsFor(normalizedModelName).get(identifier.lid);
 
     return !!internalModel && internalModel.isLoaded();
   },
 
+  _getInternalModelForIdentifier(identifier) {
+    return this._internalModelsFor(identifier.type).get(identifier.lid);
+  },
+
   // directly get an internal model from ID map if it is there, without doing any
   // processing
-  _getInternalModelForId(modelName, id, clientId) {
-    let internalModel;
-    if (clientId) {
-      internalModel = this._newlyCreatedModelsFor(modelName).get(clientId);
-    }
+  _getInternalModelForId(type, id, lid) {
+    let identifier = recordIdentifierFor(this, { type, id, lid });
```

Comment:
We are now doing two lookups per IM vs one, that seems a bit unnecessary to me. I would be excited to talk about what the benefits of the cache per identifier thats stable vs keying off of type/lid are

Reply:
previously, to lookup an internalModel, we needed to:

* access `store._identityMap._map[type]._idForModel[id]` (path length 5, two dictionaries and two classed involved)

If we missed, we would fallback to:

* `store._newlyCreated._map[type]._idForModel[clientId]` (path length 5, two dictionaries and two classes involved)

We would then use `internalModel` as a key for many things, including:

* record
* recordData
* network state

----

In the new system, there is never a fallback for newly created records. All lookups take the following journey:

* `<store>.typeMap[type].lid[lid]` (path length 4, dictionaries 2, no classes involved)

In the case wherein this is our first time ever seeing this identifier and/or no `lid` yet exists or was provided, we must fallback to `id` and then generate an `lid` if necessary. This takes an additional lookup:

* `typeCache.id[id]` (path length 2, dictionaries 1, no classes)

There is an optimization here that has not been made yet if we RFC a design choice (if not already part of RecordData RFC?): if `lid` MUST be unique across types, then we can have a single level cache with an optional `type + id` fallback cache.

*  `lidCache[lid]` (path length 1, dictionaries 1, no classes)
* `<store>.typeMap[type].id[id]` (path length 4, dictionaries 2, no classes involved)

Regardless, from here it is a single lookup (path length 1, 1 dictionary) to access `_internalModel`.

**Takeaways**

* The implemented strategy is roughly equivalent for accessing `_internalModel`, but in a common case does *slightly* regress.
* There exists an optimization for this strategy that leads us to doing *slightly* better instead.

So where, other than improving method signatures and a clearer separation of concerns internally (identifier now stores identity info instead of internalModel), is the big win / reason to do this?

**It's in the very next lookup!**

Right now, `InternalModel` is the key for `RecordData` and `Record`, but this cannot continue to be with the landing of the `modelFactoryFor` RFC, in which `recordData` is handed directly to the `record` and no `InternalModel` is. While we could continue to spin up `InternalModel`s just to serve as the cache for these things, it seems better to instead create some clear boundaries. Using `Identifier` as the key instead, we would then create the following mappings, reducing overall lookups and _reducing the memory footprint by moving from deeply nested object graphs to WeakMaps [1]:_

* `Identifier => RecordData`
* `Record  => RecordData`
* `Identifier => Record`

In `DEBUG` mode, we would symbolicate these object relationships and provide access to the `Identifier` map via a symbol on `Store` so that the entire data store would be discoverable and debuggable in the console without leaking the APIs for abuse.

[1] _keying by `lid` or by `type + id` creates higher memory stress due to the way memory blocks are allocated for the objects involved. As the cache grows, especially as a 2+ level cache grows, this becomes slower and much less ideal. Keying via a WeakMap has roughly identical lookup speed to a dictionary but without this memory overhead_